### ISSUE TITLE
Accessibility of assets field UI in entries

### ISSUE THREAD

**Issue by ajanec01** [craftcms/cms#7366]
#### Basic UI 

- <s>The add an asset and upload files components would benefit from `role="group"` or `<fieldset> `; the name of the asset field is not announced when the focus is moved to either of the two buttons</s> Fixed
![grouping-assets](https://user-images.githubusercontent.com/47136246/103952458-5ae66880-5138-11eb-9579-701d98feaf0b.png)
- <s>Poor keyboard focus indication on all components</s> Fixed
- <s>The remove asset links are not keyboard accessible (should be buttons-) they don’t have `href`; they should also have the programmatically determined link context, or add a screen reader only text</s> Fixed
![assets-display](https://user-images.githubusercontent.com/47136246/103955558-73598180-513e-11eb-9e85-4d12c5819535.png)
- Not possible to move the assets up or down using the keyboard only
- Only one of the assets can be navigated into using the Tab- the rest can be accessed using the arrow keys which is confusing
- The assets triggers need to be buttons 
- Not possible to open the asset info dialog using the keyboard only
![asset-information-dialog](https://user-images.githubusercontent.com/47136246/103955946-5ec9b900-513f-11eb-923e-6fd797a07634.png)
- The asset info dialog has accessibility issues covered in #7297 and in #7308 
- The additional info trigger is not keyboard accessible; it also needs accessible role and state (consider `<details>` and `<summary>`)
![additional-info](https://user-images.githubusercontent.com/47136246/103956486-82d9ca00-5140-11eb-889a-69dbaf3497f3.png)

#### Assets modal dialog

- the modal dialog with assets does not have acc name and role + the focus is not trapped- details in #7297 
![assets-modal](https://user-images.githubusercontent.com/47136246/103952994-3939b100-5139-11eb-9778-2d323134457e.png)
- Couple of issues in the navigation panel:
  - Whenever the trigger button is activated using the keyboard, the focus moves away from the trigger; need to do shift + tab to come back
![expandable-assets](https://user-images.githubusercontent.com/47136246/103953050-51113500-5139-11eb-9130-3abc1010b628.png)
  - Only one of the folders receives keyboard focus, a click on either of them adds `tabindex`- in short, possible to access only one folder using the keyboard only
  - Not sure if assets folder structure should be wrapped inside a `<nav>`; tree view seems more appropriate as it indicates the number of subfolders so it would be easier to navigate; worth having a look at tree view  [example 1 ](https://www.w3.org/TR/wai-aria-practices-1.2/examples/treeview/treeview-1/treeview-1a.html )and [example 2](https://www.w3.org/TR/wai-aria-practices-1.2/examples/treeview/treeview-1/treeview-1b.html ); if applied correctly, should resolve all of the navigation issues
- Use `<input type="search" />` or `role="searchbox"`  for the searchbox
- The clear search button is not keyboard accessible and is missing the acc role
- The sort by button could give a bit more context to screen reader users, consider using a visually hidden text
- Custom fields for assets are announced only once because they have the same id as the value of `aria-controls` (menu.Garnish763093901-option-0)
![focusss-loss](https://user-images.githubusercontent.com/47136246/103954073-57081580-513b-11eb-8d4c-521cc9a81d12.png)
- There are two listboxes inside the filter (for values and for ascending/descending); using one `listbox` and grouping content would make it more understandable- see example [aria listbox](https://www.w3.org/TR/wai-aria-practices-1.2/examples/listbox/listbox-grouped.html)
- The display buttons (table/ thumbnails) function as radio buttons so it would be better to make them a group of radio buttons with a legend (at the moment they don’t convey the selected state)
- Add `scope` to row headers (nitpicking- sr works fine without)
![row-scope](https://user-images.githubusercontent.com/47136246/103954746-c599a300-513c-11eb-99ab-299924a0a2e4.png)
- It’s a bit confusing to choose an asset in the table view, there is no indication to sr users that a row has been selected, maybe a little explanation beforehand, or a checkbox in a column next to the asset's name
- Not possible to sort assets by column headers (using keyboard only)- they are also missing `aria-sort`; similar solution in example 2 for [aria grid ](https://www.w3.org/TR/wai-aria-practices-1.2/examples/grid/dataGrids.html)
- Thumbnails can be navigated using arrow keys, they resemble radio buttons but do not have the a11y features- confusing to sr users
- Not possible to select another thumbnail using the keyboard only after you’ve added one already; 
- Adding an asset could convey status to sr users- a visually hidden text with `aria-live="polite"` ?

### Additional info

- Craft version: 3.5.17.1

**Comment by gcamacho079** [comment]
### Issue Updates
In addition to the issues previously marked “Fixed,” these issues have been addressed or rendered stale by UI changes:

#### Keyboard:
- Assets can be moved up/down using buttons in the “Actions” disclosure menu, which can be activated via keyboard
- All asset controls in the disclosure menu can be navigated to using the keyboard
- In the “Add an asset” dialog:
    - All sources in the sidebar are in the tab order
    - The clear search button is accessible via the keyboard
    - Column headers can be sorted via keyboard
    - Multiple assets can be selected via the checkboxes in the tab order

#### Screen reader:

- In the “Add an asset” dialog:
    - The dialog has an accessible name and traps focus
    - The clear search button has an accessible name
    - The “Sort by” select has a meaningful, accessible name (”Sort attribute”)
    - Issues caused by duplicate IDs and multiple listboxes in the sort attribute/direction select have been eliminated since the UI was updated to use a disclosure menu with native HTML inputs
    - Display buttons have been updated to use an “exclusive button group” pattern, and NRV is communicated
    - Table header `scope` has been added
    - A checkbox column has been added to indicate for screen reader users which rows are selected
    - Column headers include the `aria-sort` attribute
    - When assets are added or removed from the field, the entry live region announces: “Your changes have been stored.”
    
#### Unchanged:
- The search input has retained a `text` role
- For the source navigation, we’ve chosen to maintain a `nav` pattern with nested buttons as the pattern may be more intuitive to most users
